add-interactive.c

```diff
@@ -47,27 +47,20 @@ struct list_and_choose_options {
 
 struct choice {
 	struct hashmap_entry e;
-	union {
-		void (*command_fn)(void);
-		struct {
-			struct adddel index, worktree;
-		} file;
-	} u;
 	size_t prefix_length;
 	const char *name;
 };
 
-enum choice_type {
-	FILE_STAT,
-	COMMAND
+struct file_choice {
+	struct choice choice;
+	struct adddel index, worktree;
 };
 
 struct choices {
 	struct choice **choices;
 	size_t alloc, nr;
-	enum choice_type type;
 };
-#define CHOICES_INIT { NULL, 0, 0, 0 }
+#define CHOICES_INIT { NULL, 0, 0 }

```

Comment:
Oh, sorry, I missed this earlier: using `{ NULL } ` should be plenty enough: the standard allows for specifying too few values, and just fills in the rest of the struct by using NUL bytes. That way, you don't have to change the `_INIT` constant when you extend the struct.